docs/job-writing-guide.md

```diff
@@ -0,0 +1,750 @@
+no more operations left. The output of the final operation is the output of the
+Job.
+
+![Job Pipeline](/img/guide-job-pipeline.webp) For example, here's how we issue a
+GET request with a http adaptor:
+
+```js
+get('/patients');
+```
+
+The first argument to `get` is the path to request from (the configuration will
+tell the adaptor what base url to use). In this case we're passing a static
+string, but we can also pass a value from state:
+
+```js
+get(dataValue('endpoint'));
+```
+
+The `dataValue` function will read a path from state.data (in this case,
+`state.data.endpoint`) and return it.
+
+Your job code should only contain operations at the top level/scope - you should
+NOT include any other Javascript statements. We'll talk about this more in a
+minute.
+
+<!-- TODO this is half finished -->
+<details>
+<summary>More details for Javascript Developers</summary>
+An operation is actually a factory function. It takes the arguments passed in job code, traps them in a closure, and returns a function which takes state and returns state.
+
+The Job code is not actually runnable Javascript - we compile it into an array
+of promises, basically, and then execute those promises with an async reducer.
+Compilation affects the top level operations (which get moved into that array of
+promises), but nothing else in the code.
+
+</details>
+
+## Callbacks and fn()
+
+Many operations give you access to a callback function.
+
+Callbacks will be invoked with state, will run whatever code you like, and must
+return the next state. Usually your callback will be invoked as the very last
+step in an operation.
+
+This is useful for intercepting and manipulating the return value of a given
+operation.
+
+<details>
+<summary>What is a callback?</summary>
+A callback is a common pattern in Javascript.
+
+It's kind of hard to understand in the abstract: a callback is a function which
+you pass into some other function, to be invoked by that function at a
+particular time.
+
+It's best explained with an example. All Javascript arrays have a function
+called `map`, which takes a single callback argument.
+
+Array.map will iterate over every item in the array, invoke your callback
+function with it, save the result toa new array, and when it's finished, it will
+return that array.
+
+```js
+const array = ['a', 'b', 'c'];
+const result = array.map(item => {
+  return item.toUpperCase();
+});
+console.log(array); // ['a', 'b', 'c'];
+console.log(result); // ['A', 'B', 'C'];
+```
+
+Because functions are data in Javascript, we can we-write that code like this
+(which might be a bit more readable)
+
+```js
+const array = ['a', 'b', 'c'];
+const upperCase = item => {
+  return item.toUpperCase();
+};
+const result = array.map(upperCase);
+console.log(array); // ['a', 'b', 'c'];
+console.log(result); // ['A', 'B', 'C'];
+```
+
+</details>
+
+The `fn()` function, for example, ONLY allows you define a callback. This is
```

Comment:
@josephjclark should we introduce also `common` here? Nowhere in docs so far do we mention that this is a thing... maybe we could also link to [its Adaptor page](https://docs.openfn.org/adaptors/packages/common-docs)?

Reply:
The thing with common is that it should probably be mostly invisible. I think we just need to work out how to list the common functions inside the other adaptors  (which I've raised as #440)